test/e2e/node/kubelet.go

```diff
@@ -385,6 +386,102 @@ var _ = SIGDescribe("kubelet", func() {
+			containerName := "test-removed"
+			image := imageutils.GetE2EImage(imageutils.BusyBox)
+			podTemplate := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: podName,
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyOnFailure,
+					Containers: []v1.Container{
+						{
+							Name:  containerName,
+							Image: image,
+							Command: []string{
+								"/bin/sh", "-c", "exit 1",
+							},
+						},
+					},
+				},
+			}
+
+			ginkgo.DeferCleanup(func(ctx context.Context) error {
+				ginkgo.By("deleting the pod")
+				return c.CoreV1().Pods(ns).Delete(ctx, podName, metav1.DeleteOptions{})
+			})
+
+			ginkgo.By("submitting the pod to kubernetes")
+			commonPod, err := c.CoreV1().Pods(ns).Create(ctx, podTemplate, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			err = e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, commonPod.Name, ns)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("check the pod's restart count")
+			var createdPod *v1.Pod
+			eachPoll, timeout := 20*time.Second, 10*time.Minute
+			err = wait.PollUntilContextTimeout(ctx, eachPoll, timeout, true, func(ctx context.Context) (bool, error) {
+				createdPod, err = c.CoreV1().Pods(ns).Get(ctx, commonPod.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				containerStatus := e2epod.FindContainerStatusInPod(createdPod, containerName)
+				if containerStatus == nil {
+					framework.ExpectNoError(fmt.Errorf("%s container doesn't exist in %s pod", containerName, podName))
+				}
+				// wait for container restarts 5 times
+				// backoff time: 10s, 20s, 40s, 1m20s, 2m40s, 5m
+				if containerStatus != nil && containerStatus.RestartCount >= 5 {
+					framework.Logf("%s container's restart count %d has been greater than 5", containerName, containerStatus.RestartCount)
+					return true, nil
+				}
+				return false, nil // retry
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("check the number of termination message log files")
+			nodeIP, err := getHostExternalAddress(ctx, c, createdPod)
+			framework.ExpectNoError(err)
+
+			terminationLogDir := filepath.Join(framework.TestContext.KubeletRootDir, "pods", string(createdPod.UID), "containers", containerName)
+			cmd := fmt.Sprintf("sudo ls %s | wc -l", terminationLogDir)
+
+			var fileCount int
+			err = wait.PollUntilContextTimeout(ctx, eachPoll, timeout, true, func(ctx context.Context) (bool, error) {
+				result, err := e2essh.NodeExec(ctx, nodeIP, cmd, framework.TestContext.Provider)
+				framework.ExpectNoError(err)
+				e2essh.LogResult(result)
+				ok := (result.Code == 0 && len(result.Stdout) > 0 && len(result.Stderr) == 0)
+				if !ok { // keep trying
+					return false, nil
+				}
+
+				tidyOut := strings.ReplaceAll(strings.ReplaceAll(result.Stdout, " ", ""), "\n", "")
+				fileCount, err = strconv.Atoi(tidyOut)
+				if err != nil {
+					framework.ExpectNoError(fmt.Errorf("failed to convert string %s to integer, err: %w", result.Stdout, err))
```

Comment:
I haven't seen this pattern before. I looked up quickly and it seem this is being used. I would think that the alternative that check for `err` and has extra string would be preferable.

@pohly do you have an opinion?

Reply:
This seems unnecessarily complex. The whole poing of `ExpectNoError` is that you don't need to add all those `if` checks. So instead of `if ... framework.ExpectNoError`, use:
```
framework.ExpectNoError(err, "convert string %s to integer", result.Stdout)
```

Reply:
I wonder if some linter may detect `ExpectNoError(fmt.Errorf` pattern? There are a few cases like this in codebase today

Reply:
@pohly yes, this is my mistake. I did not use this method according to the documentation. I will modify this code. :)

Reply:
To be fair, the documentation was poor. Let's make it better: https://github.com/kubernetes/kubernetes/pull/127558